p2p/discovery.py

```diff
@@ -88,9 +88,10 @@ def __init__(self, privkey: datatypes.PrivateKey, address: kademlia.Address,
         self.cancel_token = CancelToken('DiscoveryProtocol')
 
     async def lookup_random(self, cancel_token: CancelToken) -> List[kademlia.Node]:
-        node_id = random.randint(0, kademlia.k_max_node_id)
-        token_chain = self.cancel_token.chain(cancel_token)
-        return await self.kademlia.lookup(node_id, token_chain)
+        return await self.kademlia.lookup_random(self.cancel_token.chain(cancel_token))
```

Comment:
I'm trying to understand where `wait_with_token` needs to be used and doesn't.  I see that this passes in the chained tokens, however, since this is an `await` I would have expected this to need to be:

```python
return await wait_with_token(self.kademlia.lookup_random(self.cancel_token.chain(cancel_token)), token=cancel_token)
# or maybe?
return await wait_with_token(self.kademlia.lookup_random(self.cancel_token.chain(cancel_token)), token=self.cancel_token.chain(cancel_token))
```

This strikes me as something that needs to be written down somewhere in developer documentation (or maybe even just extracted into a stand-alone library and documented there.)

Reply:
That's a good point. Basically, CancelTokens are supposed to be passed on all the way through **our** async APIs until we end up calling any of python's asyncio APIs (i.e. as deep as possible in the call stack), at which point we have to use `wait_with_token()` so that the futures (created by `asyncio.wait()` and similar APIs) are cancelled when the token is triggered. IOW, whenever we use `asyncio.wait()` & co, we should use `wait_with_token()`. Does that make sense?

Reply:
Created https://github.com/ethereum/py-evm/issues/449 for this